pkg/sidecar-controller/snapshot_controller.go

```diff
@@ -564,9 +595,71 @@ func (ctrl *csiSnapshotSideCarController) shouldDelete(content *crdv1.VolumeSnap
 	if content.Spec.Source.SnapshotHandle != nil && content.Spec.VolumeSnapshotRef.UID == "" {
 		return true
 	}
-	// 2) shouldDelete returns true if AnnVolumeSnapshotBeingDeleted annotation is set
+
+	// NOTE(xyang): Handle create snapshot timeout
+	// 2) shouldDelete returns false if AnnVolumeSnapshotBeingCreated
+	// annotation is set. This indicates a CreateSnapshot CSI RPC has
+	// not responded with success or failure.
+	// We need to keep waiting for a response from the CSI driver.
+	if metav1.HasAnnotation(content.ObjectMeta, utils.AnnVolumeSnapshotBeingCreated) {
```

Comment:
The deletion timestamp would have already been set on the content by that time.  So even if we don't add "AnnVolumeSnapshotBeingDeleted", it is still being deleted in fact.  So I don't know if it helps by delaying that annotation.

Reply:
Are you suggesting to skip the deletion of VolumeSnapshotContent until later in the snapshot controller?  The Content deletion is done in syncSnapshot.  If we delay the deletion of Content there, when will it be triggered again?  I think that could delay the deletion of the snapshot which is not necessarily good user experience.

Reply:
I did some testing with this PR.  Try to delete the snapshot before create snapshot is completed (added a sleep in the driver).  I found that the PVC finalizer won't be removed in that case because the physical snapshot, the content, and finalizer on the snapshot are all deleted before the common controller has a chance to do another syncSnapshot.

So I fixed it in the latest PR by adding a check in the common controller to see if the PVC is still being used by the snapshot as a source.  If it is, then the common controller will skip processing the deletion stamp and wait until later.  This works.  

Note: I tried earlier to check the "AnnVolumeSnapshotBeingCreated" but that does not fix this problem.

I think we should re-evaluate when we remove the PVC finalizer and what does it mean by "in-use".  Currently we check if PVC is a source for a snapshot and whether that snapshot is readyToUse.  This has problems.  If create snapshot fails (even with a terminating failure), readyToUse is false.  If snapshot is already cut but it is being uploaded, readyToUse is false but we should not need to keep the finalizer on PVC any more.  I think we should remove the PVC finalizer when either a creation timestamp is set or when a terminating failure has returned.

Reply:
This fix for removing PVC finalizer will be submitted in a separate PR.